### DISCUSSION TITLE
How to build plugins to support CSS-in-JS with evaluation phase

### DISCUSSION THREAD

**Discussion by chentsulin** [swc-project/swc#6594] [Q&A] [answered]
Some Zero Runtime CSS-in-JS solutions (for example: [Linaria](https://github.com/callstack/linaria) and unreleased Stylex from Meta) now highly rely on [JS runtime evaluation phase](https://github.com/callstack/linaria/blob/master/packages/babel/src/transform-stages/2-eval.ts) using Babel.

They need to evaluate JS expressions and then replace `1 + 1` with `2` or even replace `Math.min(1, 2)` function call with `1`.

Is there anything we can do to migrate this kind of tools to SWC? Using SWC with these tools hurts build time performance a lot due to running the Babel part inevitably.

**Comment by kwonoj** [comment, marked as answer]
I do not believe there is a way to dynamically evaluate values in the current plugin implementation. After all, SWC runs in a rust's host runtime context and plugin also runs in its wasm runner, not in the JS context.

**Comment by kdy1** [comment]
If you want to evaluate expressions in compile time, see https://rustdoc.swc.rs/swc_ecma_minifier/eval/struct.Evaluator.html

**Reply by kdy1** [reply]
But it does not support evaluating js. It's about analyzing statically analyzable codes.

**Reply by kwonoj** [reply]
afaik most of css-in-js eval is runtime evaluation unfortunately (i.e https://github.com/callstack/linaria/blob/ac2d341eef68b9fce995014735dc591db965caf6/packages/babel/src/transform-stages/2-eval.ts#L32)
